### PR TITLE
Add login test that runs against Anon and Authorized table.

### PR DESCRIPTION
**Pull request by aziel**
This adds testing Login against Anonymous and Authorized tables for the WinStore app.

### CODE REVIEW

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs [hunk 1]
```diff
@@ -36,10 +43,140 @@ class LoginTests
...
+        }
+
+        public static async Task<string> TestCRUD(string tableName, bool tableRequiresAuthentication, bool userIsAuthenticated)
+        {
+            bool crudShouldWork = (tableRequiresAuthentication && userIsAuthenticated) || !tableRequiresAuthentication;
+            var user = client.CurrentUser;
+            var table = client.GetTable(tableName);
+            var item = new JObject();
+            item.Add("name", "John Doe");
+            int id;
```
Comment (phvannor) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: why is Id an int?  New tables are all string based.

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs [hunk 2]
```diff
@@ -36,10 +43,140 @@ class LoginTests
...
+            catch (MobileServiceInvalidOperationException e)
+            {
+                ValidateExpectedFailure("Lookup", tableName, crudShouldWork, tableRequiresAuthentication, e);
+            }
+
+            // Attempt to read data with filter
+            try
+            {
+                JToken items = null;
+                if (await IsNetRuntime(queryParameters))
```
Comment (phvannor) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: This worries me, as refresh uses $filter=id eq 'id', if this is true refresh is broken on .Net?

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs [hunk 3]
```diff
@@ -36,10 +43,140 @@ class LoginTests
...
+            }
+
+            return "Test complete";
+        }
+
+        public static async Task<bool> IsNetRuntime(Dictionary<string,string> queryParams)
+        {
+            try
+            {
+                JToken runtimeInfo = await client.InvokeApiAsync("runtimeinfo", HttpMethod.Get, queryParams);
```
Comment (phvannor) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: This doesn't match the pattern of filtering we use elsewhere, in fact the test suite should already know all of this and be able to provide it globally, not every test determining it.

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs [hunk 4]
```diff
@@ -36,10 +43,140 @@ class LoginTests
             {
                 user = await client.LoginAsync(provider, useSingleSignOn);
             }
+            await TestCRUD("Public", false, true);
+            await TestCRUD("Authorized", true, true);
+            await TestLogoutAsync();
+            await TestCRUD("Public", false, false);
+            await TestCRUD("Authorized", true, false);
```
Comment (phvannor) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: you'll still want a logout post the test ends, or else that will influence other tests in unpredictable ways

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs [hunk 5]
```diff
@@ -36,10 +43,140 @@ class LoginTests
             {
                 user = await client.LoginAsync(provider, useSingleSignOn);
             }
+            await TestCRUD("Public", false, true);
+            await TestCRUD("Authorized", true, true);
+            await TestLogoutAsync();
+            await TestCRUD("Public", false, false);
```
Comment (phvannor) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: I'm not sure why we run this again post logout?  It seems overkill to run another suite of table tests

Comment (aziel) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: Wouldn't this assert that Logout actually logs out?

/iPhone

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs (hunk 1 above)
Comment (aziel) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: Changed to string.

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs (hunk 2 above)
Comment (aziel) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: I brought this over from David's old tests, when case mattered I guess.
I tested it with both Id and id
https://azielprodbay2.azurewebsites.net/tables/Public?$filter=id%20eq%20%271%27&zumo-api-version=2.0.0

Works, so I can remove the IsNetRuntime check.

**Comment by aziel** [comment]
I've updated the PR per your comments

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs [hunk 6]
```diff
@@ -36,10 +43,119 @@ class LoginTests
...
+            return "Test complete";
+        }
+
+        private static void ValidateExpectedFailure(string attemptedAction, string tableName, bool crudShouldWork, bool tableRequiresAuth, MobileServiceInvalidOperationException outerException)
+        {
+            if (crudShouldWork)
+            {
+                if (!tableRequiresAuth)
+                {
+                    throw new Exception(string.Format("Anonymous {0} to unprotected table {1} should have succeeded. Instead, found Status Code: {2}. Reason: {3}.", attemptedAction, tableName, outerException.Response.StatusCode, outerException.Response.ReasonPhrase));
```
Comment (fabiocav) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: Would be nice to refactor this to build the exception message portion that varies and avoid the duplicate code throwing the exception.

Comment (fabiocav) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: Even though we're building the exception message with data from the original exception, would still be nice to pass that exception as the inner exception when throwing.

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs [hunk 7]
```diff
@@ -36,10 +43,119 @@ class LoginTests
...
             return string.Format("UserId: {0} Token: {1}", user.UserId, user.MobileServiceAuthenticationToken);
         }
 
+        public static async Task<string> TestLogoutAsync()
+        {
+            await client.LogoutAsync();
+            return string.Format("Logged out. Current logged-in client: {0}", client.CurrentUser == null ? "<<NULL>>" : client.CurrentUser.UserId);
+        }
+
+        public static async Task<string> TestCRUD(string tableName, bool tableRequiresAuthentication, bool userIsAuthenticated)
```
Comment (fabiocav) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: nit: TestCRUDAsync ?

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs (hunk 5 above)
Comment (phvannor) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: Testing "Public" is not adding anything, only need to run an Authorized one.  Really only need to do 1 call, not the whoel CRUD set, but we can live with just authorized running again.

**Comment by phvannor** [comment]
Is there a reason we didn't try to do the "cheap" client directed login flow tests as part of this as well?

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs (hunk 5 above)
Comment (aziel) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: I see what you mean, I'm removing the repeated Public test after logout.

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs (hunk 6 above)
Comment (aziel) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: Both done.

File: e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs (hunk 7 above)
Comment (aziel) [inline e2etest/test/Microsoft.WindowsAzure.MobileServices.WindowsStore.Test/PlatformTests/LoginTests.cs]: Done

**Comment by aziel** [comment]
I'll add client directed login flow later.  I still have to adjust login tests for other projects.  This is just for winstore.

**Comment by aziel** [comment]
Updated per Fabio and Phil's comments again.  Will add client flow tests in a future PR

**Pull request merged**